Fix e2e tests for providers without SSH to the nodes

sttts opened the pull request:
A number of e2e tests which use SSH are skipped for providers other than gce, gke and aws. This patch does the same for 
- the kubeproxy tests (only for the probes which actually use SSH, the others are kept running)
- "should release NodePorts on delete" in service.go
- "should test privileged pod" in privileged.go.

Moreover, the provider list of those which support SSH is factored out into `providersWithSSH`.

Fixes https://github.com/mesosphere/kubernetes-mesos/issues/500

sttts committed:
Refactor kubeproxy tests to run with and without provider SSH support

All other e2e tests which use SSH are skipped for providers other than gce,
gke and aws. This patch does the same for the kube-proxy test.

If not SSH support is available for the provider, the test will still run,
but use less probes.

sttts committed:
Add missing guards around SSH based e2e tests

All other e2e tests which use SSH are skipped for providers other than gce,
gke and aws. This patch does the same for

- "should release NodePorts on delete" in service.go
- "should test privileged pod" in privileged.go.

sttts commented:
/cc @s-urbaniak for internal review

s-urbaniak commented on test/e2e/kubeproxy.go:
```diff
@@ -417,13 +428,32 @@ func (config *KubeProxyTestConfig) setup() {
+		}
+	}
+	if len(config.nodes) == 0 {
+		Failf("No nodes with externally reachable address found")
```
`Failf` implies the whole test must fail, is this true?

sttts replied:
yes, fatal error.

s-urbaniak commented on test/e2e/service.go:
```diff
@@ -532,14 +532,17 @@ var _ = Describe("Services", func() {
+			}
+			cmd := fmt.Sprintf(`test -n "$(ss -ant46 'sport = :%d' | tail -n +2 | grep LISTEN)"`, nodePort)
+			_, _, code, err := SSH(cmd, hosts[0], testContext.Provider)
+			if code != 0 {
```
Can `code == 0 && err != nil` happen? if yes, we should check for `err != nil`?

sttts replied:
probably. But I this is kind of out of scope of this PR. The code is the old code, just indented.

s-urbaniak commented:
LGTM

jdef commented:
rather than copy/pasting `providerIs("gce", "gke", "aws")` in several places, is it worth adding a top-level func `ProviderSupportsSSH() bool` that contains this logic?

sttts commented:
@jdef I like that idea.

sttts committed:
Factor out providers which support SSH in e2e tests

sttts commented:
Rebased and squashed.

sttts commented:
@s-urbaniak PTAL, rebased and squashed.

s-urbaniak commented:
LGTM

sttts commented:
@ArtfulCoder I have assigned this to you for a review as you wrote most of the kubeproxy e2e tests.

ArtfulCoder commented on test/e2e/kubeproxy.go:
```diff
@@ -60,6 +61,7 @@ type KubeProxyTestConfig struct {
 	nodePortService     *api.Service
 	loadBalancerService *api.Service
 	nodes               []string
+	enableSSHTests      bool
```
where is this variable used ?

sttts replied:
Left over from a refactoring. enableSSHTests must go.

ArtfulCoder commented on test/e2e/kubeproxy.go:
```diff
@@ -38,7 +40,6 @@ import (
 const (
 	endpointHttpPort        = 8080
 	endpointUdpPort         = 8081
-	endpointHostPort        = 8082
```
why was this removed ?

also, if you have good reason to remove it, we should do it in a separate PR..

sttts replied:
It had been in its own commit before squashing. Will extract it again.

About the why: it doesn't seem to be used.

ArtfulCoder replied:
it is required to  ensure that the pods dont share the same host.

sttts replied:
Oh, then this deserves a comment ;-)

sttts replied:
Would you mind to set the NodeName correctly for each pod instead of this "port hack". Constant host ports are not nice.

ArtfulCoder replied:
its a fair suggestion. I would keep the hostPort and do the NodeName in a separate PR.
Lets not combine things in the same PR.

ArtfulCoder commented on test/e2e/kubeproxy.go:
```diff
@@ -225,13 +228,17 @@ func (config *KubeProxyTestConfig) dialFromContainer(protocol, containerIP, targ
-	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Could not unmarshal curl response: %s", stdout))
-	hostNamesMap := array2map(output["responses"])
-	Expect(len(hostNamesMap)).To(BeNumerically("==", expectedCount), fmt.Sprintf("Response was:%v", output))
+	if providerIs(providersWithSSH...) {
```
I think we should use kubectl exec into a net=host pod and run the shell commands to work around the no-ssh-available problem.
For purposes of this PR, I would just run the entire test  for providerWithSSH, otherwise skip the whole test.
We are skipping pretty much everything other than the LoadBalancer portion..

ArtfulCoder replied:
slightly off-topic:
In general, we want to use this container to run shell commands on the node.
 https://github.com/kubernetes/kubernetes/tree/master/contrib/for-tests/netexec 
We can use kubectl to hit this container using kubectl exec..

sttts replied:
I like the idea of a net=host pod. net=host pods are disabled though AFAIK for every other pod config source than file. This can be overwritten by a kubelet flag. So you have to configure your cluster for net=host support.

Thoughts?

ArtfulCoder replied:
yes, i remember now 
i had worked on that file-based net=host pod feature :)
I guess, there isnt an easy way to do that yet..

ArtfulCoder replied:
you wont need the if now

ArtfulCoder commented on test/e2e/kubeproxy.go:
```diff
@@ -417,13 +423,32 @@ func (config *KubeProxyTestConfig) setup() {
+			config.nodes = append(config.nodes, strings.TrimSuffix(h, ":22"))
+		}
+	} else {
+		// fall back to legacy IP
```
can this logic be in the NodeSSHHosts function ?

sttts replied:
It does not make sense in general I think, only for this test. But I removed this now, following your comment above to skip the whole test. I agree a net=host pod is the way to go.

sttts commented:
@ArtfulCoder addressed all your comments.

What do you think about the net=host question, i.e. that it requires either a static pod or a different kubelet configuration?

ArtfulCoder commented on test/e2e/kubeproxy.go:
```diff
@@ -242,13 +248,17 @@ func (config *KubeProxyTestConfig) dialFromNode(protocol, targetIP string, targe
-	By(fmt.Sprintf("Dialing from node. command:%s", forLoop))
-	stdout := config.ssh(forLoop)
-	Expect(strconv.Atoi(strings.TrimSpace(stdout))).To(BeNumerically("==", expectedCount))
+	if providerIs(providersWithSSH...) {
```
you wont need the if now..

ArtfulCoder commented on test/e2e/kubeproxy.go:
```diff
@@ -284,11 +294,6 @@ func (config *KubeProxyTestConfig) createNetShellPodSpec(podName string) *api.Po
 							ContainerPort: endpointUdpPort,
 						},
-						{
-							Name:          "host",
```
need this back in

ArtfulCoder commented on test/e2e/kubeproxy.go:
```diff
@@ -420,10 +425,10 @@ func (config *KubeProxyTestConfig) setup() {
 	By("Getting ssh-able hosts")
 	hosts, err := NodeSSHHosts(config.f.Client)
 	Expect(err).NotTo(HaveOccurred())
-	if len(hosts) == 0 {
```
can we add this check back

sttts replied:
NodeSSHHosts checks this:
```
if len(hosts) != len(nodelist.Items) {
```

We won't run this test on an empty cluster I guess.

ArtfulCoder replied:
We won't run the test and the test will appear to be green when it was in-fact not run.
So, this check basically says that: something is wrong.

ArtfulCoder replied:
sounds good.

ArtfulCoder commented on test/e2e/kubeproxy.go:
```diff
@@ -420,10 +425,10 @@ func (config *KubeProxyTestConfig) setup() {
-		Failf("No ssh-able nodes")
+	config.nodes = make([]string, len(hosts))
+	for _, h := range hosts {
+		config.nodes = append(config.nodes, strings.TrimSuffix(h, ":22"))
```
can we append :22 when we actually make the ssh call

sttts replied:
we do, see config.ssh

ArtfulCoder replied:
yes, thanks

sttts committed:
Enlarge port range for mesos/docker e2e tests

sttts commented:
@ArtfulCoder ptal. I have some more patches in the pipe, e.g. NodeName based scheduling, parallel startup of the pod and a prototype of kubectl exec instead of ssh. Will create new PRs later for that.

sttts commented:
Rebased and squashed.

ArtfulCoder commented:
for the hostNetwork=true question, I have a PR in flight to allow that for non-file-based pods as well.
#15021

ArtfulCoder commented:
@sttts 
Basically, even if we cannot SSH on to nodes, we can use a pod with hostNetwork = true (post- #15021) and run shell commands from the pod.we can use privileged=true if required.(#13930)

sttts commented:
@ArtfulCoder very cool about allowing net=host by default. Let me push my port to that next week. Will also take a look at the other ssh using e2e tests. Getting rid of ssh would be great.

sttts commented:
Shippable 1.4 times out, 1.3 is green. Rebased. Let's see.
